drivers/osdp/src/osdp_pd.c

```diff
@@ -0,0 +1,576 @@
+/*
+ * Copyright (c) 2020 Siddharth Chandrasekaran <siddharth@embedjournal.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <logging/log.h>
+LOG_MODULE_DECLARE(osdp, CONFIG_OSDP_LOG_LEVEL);
+
+#include <stdlib.h>
+#include <string.h>
+
+#include "osdp_common.h"
+
+#define TAG "PD: "
+
+extern struct osdp osdp_ctx;
+extern struct osdp_cp osdp_cp_ctx;
+extern struct osdp_pd osdp_pd_ctx[];
+
+struct pd_id osdp_pd_id = {
+	.version = CONFIG_OSDP_PD_ID_VERSION,
+	.model = CONFIG_OSDP_PD_ID_MODEL,
+	.vendor_code = CONFIG_OSDP_PD_ID_VENDOR_CODE,
+	.serial_number = CONFIG_OSDP_PD_ID_SERIAL_NUMBER,
+	.firmware_version = CONFIG_OSDP_PD_ID_FIRMWARE_VERSION,
+};
+
+struct pd_cap osdp_pd_cap[] = {
+	/* Driver Implicit cababilities */
+	{
+		CAP_CHECK_CHARACTER_SUPPORT,
+		1, /* The PD supports the 16-bit CRC-16 mode */
+		0, /* N/A */
+	},
+
+	/* Configured from Kconfig */
+	{
+		CAP_CONTACT_STATUS_MONITORING,
+		CONFIG_OSDP_PD_CAP_CONTACT_STATUS_MONITORING_COMP_LEVEL,
+		CONFIG_OSDP_PD_CAP_CONTACT_STATUS_MONITORING_NUM_ITEMS,
+	},
+	{
+		CAP_OUTPUT_CONTROL,
+		CONFIG_OSDP_PD_CAP_OUTPUT_CONTROL_COMP_LEVEL,
+		CONFIG_OSDP_PD_CAP_OUTPUT_CONTROL_NUM_ITEMS,
+	},
+	{
+		CAP_READER_LED_CONTROL,
+		CONFIG_OSDP_PD_CAP_READER_LED_CONTROL_COMP_LEVEL,
+		CONFIG_OSDP_PD_CAP_READER_LED_CONTROL_NUM_ITEMS,
+	},
+	{
+		CAP_READER_AUDIBLE_OUTPUT,
+		CONFIG_OSDP_PD_CAP_READER_AUDIBLE_OUTPUT_COMP_LEVEL,
+		CONFIG_OSDP_PD_CAP_READER_AUDIBLE_OUTPUT_NUM_ITEMS,
+	},
+	{
+		CAP_READER_TEXT_OUTPUT,
+		CONFIG_OSDP_PD_CAP_READER_TEXT_OUTPUT_COMP_LEVEL,
+		CONFIG_OSDP_PD_CAP_READER_TEXT_OUTPUT_NUM_ITEMS,
+	},
+	{
+		CAP_CARD_DATA_FORMAT,
+		CONFIG_OSDP_PD_CAP_CARD_DATA_FORMAT_COMP_LEVEL,
+		CONFIG_OSDP_PD_CAP_CARD_DATA_FORMAT_NUM_ITEMS,
+	},
+	{
+		CAP_TIME_KEEPING,
+		CONFIG_OSDP_PD_CAP_TIME_KEEPING_COMP_LEVEL,
+		CONFIG_OSDP_PD_CAP_TIME_KEEPING_NUM_ITEMS,
+	},
+	{ -1, 0, 0 } /* Sentinel */
+};
+
+void pd_enqueue_command(struct osdp_pd *pd, struct osdp_cmd *cmd)
+{
+	struct osdp_cmd_queue *q = &pd->queue;
+
+	cmd->__next = NULL;
+	if (q->front == NULL) {
+		q->front = q->back = cmd;
+	} else {
+		assert(q->back);
+		q->back->__next = cmd;
+		q->back = cmd;
+	}
+}
+
+void pd_decode_command(struct osdp_pd *pd, struct osdp_cmd *reply,
+		       uint8_t *buf, int len)
+{
+	int i, ret = -1, pos = 0;
+	struct osdp_cmd *cmd;
+
+	reply->id = 0;
+	pd->cmd_id = buf[pos++];
+	len--;
+
+	switch (pd->cmd_id) {
+	case CMD_POLL:
+		reply->id = REPLY_ACK;
+		ret = 0;
+		break;
+	case CMD_LSTAT:
+		reply->id = REPLY_LSTATR;
+		ret = 0;
+		break;
+	case CMD_ISTAT:
+		reply->id = REPLY_ISTATR;
+		ret = 0;
+		break;
+	case CMD_OSTAT:
+		reply->id = REPLY_OSTATR;
+		ret = 0;
+		break;
+	case CMD_RSTAT:
+		reply->id = REPLY_RSTATR;
+		ret = 0;
+		break;
+	case CMD_ID:
+		pos++;		/* Skip reply type info. */
+		reply->id = REPLY_PDID;
+		ret = 0;
+		break;
+	case CMD_CAP:
+		pos++;		/* Skip reply type info. */
+		reply->id = REPLY_PDCAP;
+		ret = 0;
+		break;
+	case CMD_OUT:
+		if (len != 4)
```

Comment:
what is the magic constant (4) length here? Can you use a define or sizeof instead?
Please put {} even for one liners.

Reply:
Each command has a different structure and not all the bytes sent on wire is actually useful so I didn't create a dedicated structure for each of them. Those numbers are the number of bytes that each of those cases would operate on.

I know it's ugly but alternatives I could think of were worse. I am open for suggestions herre. :) On the up side, each case only has 5 to 10 lines of code, so we can just see in one window width how many byes are being consumed.

Reply:
I've addressed the magic constant issue.